File: demo-site/frontend/src/types/pen.ts
// TypeScript mirrors of backend Pydantic models (pen.py + source.py)
// Keep in sync with backend/models/pen.py

export type KeyRole = "primary" | "foreign" | "unique" | "audit" | "business_key" | "none";
export type ReviewStatus = "accepted" | "needs_review" | "rejected";
export type ProposalStatus = "pending" | "accepted" | "rejected";
export type EntityKind = "strong_entity" | "weak_entity" | "lookup_table";

export interface AttributeEvidence {
  source_id: string;
  source_columns: string[];
}

export interface Attribute {
  id: string;
  name: string;
  display_name?: string;
  pg_type: string;
  key_role: KeyRole;
  nullable: boolean;
  default?: string;
  check_constraint?: string;
  source_columns: string[];
  evidence: AttributeEvidence[];
  confidence: number;
  review_status: ReviewStatus;
}

export interface Entity {
  id: string;
  kind: EntityKind;
  name: string;
  display_name: string;
  domain?: string;        // owning business process (labels/grouping only; not a color)
  connects?: string[];    // departments this entity bridges; non-empty => seam (connection point)
  attributes: Attribute[];
  source_evidence: AttributeEvidence[];
  proposal_reason?: string;
  confidence: number;
  review_status: ReviewStatus;
}

export interface Cardinality {
  from_min: number;
  from_max: number | "many";
  to_min: number;
  to_max: number | "many";
}

export interface RelationshipPostgres {
  constraint_name: string;
  on_delete: "restrict" | "cascade" | "set_null" | "set_default" | "no_action";
  on_update: "restrict" | "cascade" | "set_null" | "set_default" | "no_action";
}

export interface RelationshipEndpoint {
  entity_id: string;
  attribute_id: string;
}

export interface Relationship {
  id: string;
  name?: string;
  from: RelationshipEndpoint;
  to: RelationshipEndpoint;
  cardinality: Cardinality;
  postgres?: RelationshipPostgres;
  dfd_process_id?: string;
  evidence: string[];
  confidence: number;
  review_status: ReviewStatus;
}

export interface ErdModel {
  notation: "crows_foot";
  model_level: "physical";
  entities: Entity[];
  relationships: Relationship[];
}

export interface ExternalEntity {
  id: string;
  name: string;
  description?: string;
  source: "manual" | "auto";
  source_evidence: string[];
  source_facts: string[];
}

export interface Process {
  id: string;
  number?: string;
  name: string;
  description?: string;
  user_modified: boolean;
  mapped_relationship_id?: string;
  source_evidence: string[];
  source_facts: string[];
  level_1_diagram?: DfdModel;
}

export interface DataStore {
  id: string;
  name: string;
  description?: string;
  mapped_erd_entity?: string;
  user_modified: boolean;
  source_evidence: string[];
  source_facts: string[];
}

export interface DataFlow {
  id: string;
  from: string;
  to: string;
  data_name?: string;
  mapped_relationship_id?: string;
  user_modified?: boolean;
  mapped_erd_entities: string[];
  mapped_erd_attributes: string[];
  source_evidence: string[];
  source_facts: string[];
}

export interface SystemBoundary {
  id: string;
  name: string;
}

export type RuleCategory = "invariant" | "state_machine" | "gate" | "validation" | "lifecycle";
export type RuleSeverity = "constraint" | "warning" | "best_practice";

export interface BusinessRule {
  id: string;
  context?: string;
  entity_id?: string;
  relationship_id?: string;
  process_id?: string;
  title: string;
  statement: string;
  category: RuleCategory;
  condition?: string;
  enforced_at: string[];
  spec_source?: string;
  verified_by?: string;
  severity: RuleSeverity;
  status: "enforced" | "gap" | "deferred";
  review_status: ReviewStatus;
}

export type ConnectorKind = "seam" | "webhook" | "fan_out" | "middleware" | "shared_service" | "auth";

export interface Connector {
  id: string;
  name: string;
  kind: ConnectorKind;
  trigger?: string;
  effect?: string;
  from_context?: string;
  to_contexts: string[];
  contract?: string;
  enforced_at: string[];
  spec_source?: string;
  status: "wired" | "deferred";
  review_status: ReviewStatus;
}

export interface DfdModel {
  level: number;
  notation: "yourdon_coad";
  system_boundary: SystemBoundary;
  external_entities: ExternalEntity[];
  processes: Process[];
  data_stores: DataStore[];
  data_flows: DataFlow[];
  business_rules?: BusinessRule[];
  connectors?: Connector[];
}

export interface LayoutNode {
  id: string;
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface LayoutPoint {
  x: number;
  y: number;
}

export interface LayoutEdge {
  id: string;
  route: "orthogonal" | "straight" | "bezier";
  source_handle?: string;
  target_handle?: string;
  points?: LayoutPoint[];
  label_t?: number;
  label_offset?: number;
}

export interface DiagramLayout {
  nodes: LayoutNode[];
  edges: LayoutEdge[];
}

export interface Layout {
  erd: DiagramLayout;
  dfd: DiagramLayout;
  dfd_level_1: Record<string, DiagramLayout>;
}

export interface Styles {
  theme: string;
  erd: Record<string, unknown>;
  dfd: Record<string, unknown>;
}

export interface DefaultPrimaryKey {
  type: string;
  default: string;
}

export interface PostgresSettings {
  dialect: string;
  extensions: string[];
  schema_name: string;
  default_primary_key: DefaultPrimaryKey;
  add_audit_timestamps: boolean;
}

export interface ReviewProposal {
  id: string;
  proposal_type: string;
  title: string;
  question: string;
  options: string[];
  context: Record<string, unknown>;
  source_evidence: string[];
  source_facts: string[];
  confidence: number;
  status: ProposalStatus;
}

export type WarningNodeKind =
  | "entity"
  | "relationship"
  | "attribute"
  | "data_store"
  | "process"
  | "external_entity"
  | "data_flow"
  | "global";

export interface WarningEntry {
  rule_id: string;
  severity: "blocking" | "warning";
  node_kind: WarningNodeKind;
  node_id?: string | null;
  node_name?: string | null;
  message: string;
  context: Record<string, unknown>;
}

export interface ReviewModel {
  proposals: ReviewProposal[];
  warnings: WarningEntry[];
  dismissed_warnings?: string[];
}

export interface SourceRef {
  id: string;
  type: "xlsx" | "csv" | "text";
  name: string;
  raw_path?: string;
}

export interface ProjectMeta {
  id: string;
  name: string;
  revision: number;
}

export type DiagramScope =
  | { diagram: "erd"; level?: "root" }
  | { diagram: "dfd"; level: "root" }
  | { diagram: "dfd"; level: "process"; process_id: string };

export interface PenFile {
  pen_version: string;
  project: ProjectMeta;
  sources: SourceRef[];
  erd: ErdModel;
  dfd: DfdModel;
  layout: Layout;
  styles: Styles;
  postgres: PostgresSettings;
  review: ReviewModel;
}

// API response types

export interface IngestResponse {
  project_id: string;
  source_tables: unknown[];
  pen: PenFile;
}

export interface Conflict {
  type: string;
  severity: "blocking" | "warning";
  message: string;
  object_id: string;
  context: Record<string, unknown>;
}

export interface ValidateResponse {
  valid: boolean;
  conflicts: Conflict[];
}

export interface StaticAnalysisIssue {
  severity: "error" | "warning";
  category: string;
  message: string;
  object_id: string;
  object_type: "entity" | "relationship" | "attribute" | "process" | "data_store" | "external_entity" | "data_flow";
  fix_suggestion: string | null;
  context: Record<string, unknown>;
}

export interface AnalyzeResponse {
  valid: boolean;
  blocking_count: number;
  warning_count: number;
  static_issue_count: number;
  conflicts: Conflict[];
  static_issues: StaticAnalysisIssue[];
}

export interface FixSuggestion {
  op: string;
  payload: Record<string, unknown>;
  label: string;
  reason: string;
}

export interface ChatMessage {
  role: "user" | "ai";
  text: string;
  suggestions?: FixSuggestion[];
}

export interface ChatResponse {
  message: string;
  suggestions: FixSuggestion[];
}

File: demo-site/backend/models/pen.py
"""
PenFile — the canonical .pen project file model.

Structure:
  sources      — where the data came from (spreadsheet files, text statements)
  erd          — semantic ERD objects: entities, attributes, relationships
  dfd          — semantic DFD objects: external_entities, processes, data_stores, data_flows
  layout       — x/y/size positioning for ERD and DFD canvas nodes
  styles       — visual theme (never read by SQL/DBML/Mermaid exporters)
  postgres     — PostgreSQL-specific settings (dialect, extensions, PK strategy)
  review       — pending proposals and warnings

Rule: SQL exporters read erd + postgres ONLY. Layout and styles are ignored.
"""
from __future__ import annotations

import uuid
from typing import Any, Literal
from pydantic import BaseModel, Field, field_validator


# ─────────────────────────────────────────────
# Sources
# ─────────────────────────────────────────────

class SourceRef(BaseModel):
    id: str
    type: Literal["xlsx", "csv", "text"]
    name: str                    # filename or short label
    raw_path: str | None = None  # original upload path (server-side temp)


# ─────────────────────────────────────────────
# ERD — Semantic objects
# ─────────────────────────────────────────────

class AttributeEvidence(BaseModel):
    source_id: str
    source_columns: list[str] = Field(default_factory=list)


class Attribute(BaseModel):
    id: str = Field(default_factory=lambda: f"attr_{uuid.uuid4().hex[:8]}")
    name: str
    display_name: str | None = None
    pg_type: str  # uuid | text | integer | bigint | numeric | boolean | date | timestamptz | jsonb
    key_role: Literal["primary", "foreign", "unique", "audit", "business_key", "none"] = "none"
    nullable: bool = True
    default: str | None = None
    check_constraint: str | None = None  # raw SQL expression e.g. "quantity > 0"
    enum_values: list[str] | None = None  # fixed value list (rendered as CREATE TYPE ... AS ENUM)
    # evidence / metadata
    source_columns: list[str] = Field(default_factory=list)
    evidence: list[AttributeEvidence] = Field(default_factory=list)
    confidence: float = 1.0
    review_status: Literal["accepted", "needs_review", "rejected"] = "accepted"


class IndexDef(BaseModel):
    id: str = Field(default_factory=lambda: f"idx_{uuid.uuid4().hex[:8]}")
    name: str
    attribute_ids: list[str] = Field(default_factory=list)  # Attribute.id refs — survive renames
    unique: bool = False


class Entity(BaseModel):
    id: str = Field(default_factory=lambda: f"ent_{uuid.uuid4().hex[:8]}")
    kind: Literal["strong_entity", "weak_entity", "lookup_table"] = "strong_entity"
    name: str          # snake_case table name
    display_name: str  # human-readable label
    description: str | None = None  # exported as COMMENT ON TABLE
    color: str | None = None        # canvas header tint (never read by SQL exporters)
    attributes: list[Attribute] = Field(default_factory=list)
    indexes: list[IndexDef] = Field(default_factory=list)
    # evidence / metadata
    source_evidence: list[AttributeEvidence] = Field(default_factory=list)
    proposal_reason: str | None = None
    confidence: float = 1.0
    review_status: Literal["accepted", "needs_review", "rejected"] = "needs_review"


class Cardinality(BaseModel):
    from_min: int = 0        # 0 = optional, 1 = mandatory
    from_max: int | Literal["many"] = "many"
    to_min: int = 1
    to_max: int | Literal["many"] = 1


class RelationshipPostgres(BaseModel):
    constraint_name: str
    on_delete: Literal["restrict", "cascade", "set_null", "set_default", "no_action"] = "restrict"
    on_update: Literal["restrict", "cascade", "set_null", "set_default", "no_action"] = "no_action"


class RelationshipEndpoint(BaseModel):
    entity_id: str
    attribute_id: str  # the FK attribute on the "from" side or PK on the "to" side


class Relationship(BaseModel):
    id: str = Field(default_factory=lambda: f"rel_{uuid.uuid4().hex[:8]}")
    name: str | None = None   # optional verb label (e.g. "places")
    from_: RelationshipEndpoint = Field(alias="from")
    to: RelationshipEndpoint
    cardinality: Cardinality = Field(default_factory=Cardinality)
    postgres: RelationshipPostgres | None = None
    dfd_process_id: str | None = None  # DFD Process.id if auto-generated from this relationship
    # evidence / metadata
    evidence: list[str] = Field(default_factory=list)
    confidence: float = 1.0
    review_status: Literal["accepted", "needs_review", "rejected"] = "needs_review"

    model_config = {"populate_by_name": True}


class ErdModel(BaseModel):
    notation: Literal["crows_foot"] = "crows_foot"
    model_level: Literal["physical"] = "physical"
    entities: list[Entity] = Field(default_factory=list)
    relationships: list[Relationship] = Field(default_factory=list)


# ─────────────────────────────────────────────
# DFD — Semantic objects
# ─────────────────────────────────────────────

class SystemBoundary(BaseModel):
    id: str = Field(default_factory=lambda: f"boundary_{uuid.uuid4().hex[:8]}")
    name: str = "System"


class ExternalEntity(BaseModel):
    id: str = Field(default_factory=lambda: f"ext_{uuid.uuid4().hex[:8]}")
    name: str
    description: str | None = None
    source: Literal["manual", "auto"] = "manual"
    source_evidence: list[str] = Field(default_factory=list)
    source_facts: list[str] = Field(default_factory=list)


class Process(BaseModel):
    id: str = Field(default_factory=lambda: f"proc_{uuid.uuid4().hex[:8]}")
    number: str | None = None   # e.g. "1.0", "2.1"
    name: str
    description: str | None = None
    user_modified: bool = False              # True if user manually edited name/description
    mapped_relationship_id: str | None = None  # ERD Relationship.id that spawned this process
    source_evidence: list[str] = Field(default_factory=list)
    source_facts: list[str] = Field(default_factory=list)
    # Level 1: sub-processes (if this is a Level 0 process being decomposed)
    level_1_diagram: DfdModel | None = None


class DataStore(BaseModel):
    id: str = Field(default_factory=lambda: f"store_{uuid.uuid4().hex[:8]}")
    name: str
    description: str | None = None
    mapped_erd_entity: str | None = None  # ERD Entity.id (1:1 enforced by type)
    user_modified: bool = False           # True if user manually edited name/description
    source_evidence: list[str] = Field(default_factory=list)
    source_facts: list[str] = Field(default_factory=list)


class DataFlow(BaseModel):
    id: str = Field(default_factory=lambda: f"flow_{uuid.uuid4().hex[:8]}")
    from_: str = Field(alias="from")  # ID of ExternalEntity | Process | DataStore
    to: str                            # ID of ExternalEntity | Process | DataStore
    data_name: str | None = None
    mapped_relationship_id: str | None = None  # ERD Relationship.id if auto-generated
    user_modified: bool = False  # True if user manually edited name/endpoints
    # optional cross-references to ERD objects
    mapped_erd_entities: list[str] = Field(default_factory=list)   # entity IDs
    mapped_erd_attributes: list[str] = Field(default_factory=list) # attribute IDs
    source_evidence: list[str] = Field(default_factory=list)
    source_facts: list[str] = Field(default_factory=list)

    model_config = {"populate_by_name": True}


class DfdModel(BaseModel):
    level: int = 0  # 0 = context diagram, 1 = detailed
    notation: Literal["yourdon_coad"] = "yourdon_coad"
    system_boundary: SystemBoundary = Field(default_factory=SystemBoundary)
    external_entities: list[ExternalEntity] = Field(default_factory=list)
    processes: list[Process] = Field(default_factory=list)
    data_stores: list[DataStore] = Field(default_factory=list)
    data_flows: list[DataFlow] = Field(default_factory=list)


# Allow Process to self-reference DfdModel
Process.model_rebuild()


# ─────────────────────────────────────────────
# Layout
# ─────────────────────────────────────────────

class LayoutNode(BaseModel):
    id: str
    x: float = 0.0
    y: float = 0.0
    width: float = 280.0
    height: float = 180.0


class LayoutPoint(BaseModel):
    x: float
    y: float


class LayoutEdge(BaseModel):
    id: str
    route: Literal["orthogonal", "straight", "bezier"] = "orthogonal"
    source_handle: str | None = None
    target_handle: str | None = None
    points: list[LayoutPoint] = Field(default_factory=list)
    label_t: float | None = None
    label_offset: float | None = None


class DiagramLayout(BaseModel):
    nodes: list[LayoutNode] = Field(default_factory=list)
    edges: list[LayoutEdge] = Field(default_factory=list)


class Layout(BaseModel):
    erd: DiagramLayout = Field(default_factory=DiagramLayout)
    dfd: DiagramLayout = Field(default_factory=DiagramLayout)
    dfd_level_1: dict[str, DiagramLayout] = Field(default_factory=dict)
    # keyed by Process.id when that process is decomposed into Level 1


# ─────────────────────────────────────────────
# Styles
# ─────────────────────────────────────────────

class ErdEntityStyle(BaseModel):
    shape: str = "rectangle"
    header_fill: str = "#f5f5f5"
    border_radius: int = 4
    border_color: str = "#d1d5db"
    font_size: int = 13


class ErdRelationshipStyle(BaseModel):
    notation: str = "crows_foot"
    line_style: str = "solid"
    line_color: str = "#6b7280"


class ErdStyles(BaseModel):
    entity: ErdEntityStyle = Field(default_factory=ErdEntityStyle)
    primary_key_attribute: dict[str, Any] = Field(default_factory=lambda: {"prefix": "PK", "font_weight": "bold"})
    foreign_key_attribute: dict[str, Any] = Field(default_factory=lambda: {"prefix": "FK"})
    unique_attribute: dict[str, Any] = Field(default_factory=lambda: {"prefix": "UQ"})
    audit_attribute: dict[str, Any] = Field(default_factory=lambda: {"prefix": ""})
    relationship: ErdRelationshipStyle = Field(default_factory=ErdRelationshipStyle)


class DfdStyles(BaseModel):
    external_entity: dict[str, Any] = Field(default_factory=lambda: {"shape": "rectangle"})
    process: dict[str, Any] = Field(default_factory=lambda: {"shape": "rounded_rectangle"})
    data_store: dict[str, Any] = Field(default_factory=lambda: {"shape": "open_ended_rectangle"})
    data_flow: dict[str, Any] = Field(default_factory=lambda: {"shape": "arrow", "label_position": "middle"})


class Styles(BaseModel):
    theme: str = "clean_default"
    erd: ErdStyles = Field(default_factory=ErdStyles)
    dfd: DfdStyles = Field(default_factory=DfdStyles)


# ─────────────────────────────────────────────
# PostgreSQL settings
# ─────────────────────────────────────────────

class DefaultPrimaryKey(BaseModel):
    type: str = "uuid"
    default: str = "gen_random_uuid()"


class PostgresSettings(BaseModel):
    dialect: str = "postgresql"
    extensions: list[str] = Field(default_factory=lambda: ["pgcrypto"])
    schema_name: str = "public"
    default_primary_key: DefaultPrimaryKey = Field(default_factory=DefaultPrimaryKey)
    add_audit_timestamps: bool = True  # auto-add created_at / updated_at


# ─────────────────────────────────────────────
# Review / Proposals
# ─────────────────────────────────────────────

class ReviewProposal(BaseModel):
    id: str = Field(default_factory=lambda: f"prop_{uuid.uuid4().hex[:8]}")
    proposal_type: str
    # e.g. "entity_split" | "foreign_key" | "enum_to_lookup" | "add_column_from_dfd"
    title: str                       # short business-readable title
    question: str                    # the review card question (business language)
    options: list[str]               # 2-3 answer choices
    context: dict[str, Any] = Field(default_factory=dict)
    # machine-readable payload for applying the proposal
    source_evidence: list[str] = Field(default_factory=list)
    source_facts: list[str] = Field(default_factory=list)
    confidence: float = 0.75
    status: Literal["pending", "accepted", "rejected"] = "pending"


class WarningEntry(BaseModel):
    """A structural diagram-rule violation surfaced to the user.

    Produced by `generators.diagram_rules.detect_violations`. See
    `backend/generators/diagram_rules.md` for the rule catalog.
    """
    rule_id: str                              # e.g. "DFD-01"
    severity: Literal["blocking", "warning"] = "warning"
    node_kind: Literal[
        "entity", "relationship", "attribute",
        "data_store", "process", "external_entity", "data_flow",
        "global",
    ] = "global"
    node_id: str | None = None
    node_name: str | None = None
    message: str
    context: dict[str, Any] = Field(default_factory=dict)


class ReviewModel(BaseModel):
    proposals: list[ReviewProposal] = Field(default_factory=list)
    warnings: list[WarningEntry] = Field(default_factory=list)
    dismissed_warnings: list[str] = Field(default_factory=list)

    @field_validator("warnings", mode="before")
    @classmethod
    def _migrate_legacy_warnings(cls, v: Any) -> Any:
        # Pre-rules-engine projects stored warnings as list[str]. Upgrade silently.
        if isinstance(v, list) and v and all(isinstance(x, str) for x in v):
            return [
                {"rule_id": "LEGACY", "severity": "warning",
                 "node_kind": "global", "message": s}
                for s in v
            ]
        return v


# ─────────────────────────────────────────────
# Top-level PenFile
# ─────────────────────────────────────────────

class ProjectMeta(BaseModel):
    id: str = Field(default_factory=lambda: f"proj_{uuid.uuid4().hex[:8]}")
    name: str = "Untitled Project"
    revision: int = 0


class PenFile(BaseModel):
    documentType: Literal["dashbot.dfdmaker"] = "dashbot.dfdmaker"
    schemaVersion: str = "0.2"
    pen_version: str = "0.1"
    project: ProjectMeta = Field(default_factory=ProjectMeta)
    sources: list[SourceRef] = Field(default_factory=list)
    erd: ErdModel = Field(default_factory=ErdModel)
    dfd: DfdModel = Field(default_factory=DfdModel)
    layout: Layout = Field(default_factory=Layout)
    styles: Styles = Field(default_factory=Styles)
    postgres: PostgresSettings = Field(default_factory=PostgresSettings)
    review: ReviewModel = Field(default_factory=ReviewModel)
